File: packages/whisker-firebase/crashlytics/android/src/main/kotlin/rs/whisker/modules/firebasecrashlytics/FirebaseCrashlyticsModule.kt
package rs.whisker.modules.firebasecrashlytics

import com.google.firebase.crashlytics.FirebaseCrashlytics
import rs.whisker.runtime.Module
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerModule
import rs.whisker.runtime.WhiskerValue

/** A Rust error report; Crashlytics groups non-fatals by the exception class and stack. */
class RustError(name: String, reason: String) : Exception("$name: $reason")

class RustPanic(message: String) : RuntimeException(message)

@WhiskerModule
class FirebaseCrashlyticsModule : Module() {
    private val crashlytics get() = FirebaseCrashlytics.getInstance()

    override fun definition() = ModuleDefinition {
        Name("FirebaseCrashlytics")
        Function("log") { args -> attempt { crashlytics.log(requireNotNull(args.firstOrNull()?.asString())) } }
        Function("setUserId") { args -> attempt { crashlytics.setUserId(args.firstOrNull()?.asString() ?: "") } }
        Function("setCustomKey") { args ->
            attempt {
                val key = requireNotNull(args.firstOrNull()?.asString())
                when (val value = args.getOrNull(1)) {
                    is WhiskerValue.Str -> crashlytics.setCustomKey(key, value.value)
                    is WhiskerValue.Int -> crashlytics.setCustomKey(key, value.value)
                    is WhiskerValue.Float -> crashlytics.setCustomKey(key, value.value)
                    is WhiskerValue.Bool -> crashlytics.setCustomKey(key, value.value)
                    else -> throw IllegalArgumentException("Unsupported custom key value")
                }
            }
        }
        Function("recordError") { args ->
            attempt {
                val report = requireNotNull(args.firstOrNull() as? WhiskerValue.Map).value
                val error = RustError(
                    requireNotNull(report["name"]?.asString()),
                    requireNotNull(report["reason"]?.asString()),
                )
                val frames = stackTrace(report)
                if (frames.isNotEmpty()) error.stackTrace = frames
                crashlytics.recordException(error)
            }
        }
        Function("recordPanic") { args ->
            // recordException writes on a background executor that the abort would cut short;
            // the uncaught-exception handler records synchronously and ends the process.
            val report = (args.firstOrNull() as? WhiskerValue.Map)?.value.orEmpty()
            val panic = RustPanic(report["reason"]?.asString() ?: "Rust panic")
            val frames = stackTrace(report)
            if (frames.isNotEmpty()) panic.stackTrace = frames
            val thread = Thread.currentThread()
            thread.uncaughtExceptionHandler?.uncaughtException(thread, panic)
            WhiskerValue.Map(mapOf("value" to WhiskerValue.Null))
        }
        Function("isCollectionEnabled") { _ ->
            try { success(WhiskerValue.Bool(crashlytics.isCrashlyticsCollectionEnabled)) } catch (error: Exception) { failure(error) }
        }
        Function("setCollectionEnabled") { args ->
            attempt { crashlytics.isCrashlyticsCollectionEnabled = requireNotNull(args.firstOrNull()?.asBool()) }
        }
        AsyncFunction("checkForUnsentReports") { _, promise ->
            try {
                crashlytics.checkForUnsentReports().addOnCompleteListener { task ->
                    promise.resolve(if (task.isSuccessful) success(WhiskerValue.Bool(task.result == true)) else failure(task.exception))
                }
            } catch (error: Exception) { promise.resolve(failure(error)) }
        }
        Function("sendUnsentReports") { _ -> attempt { crashlytics.sendUnsentReports() } }
        Function("deleteUnsentReports") { _ -> attempt { crashlytics.deleteUnsentReports() } }
        Function("didCrashOnPreviousExecution") { _ ->
            try { success(WhiskerValue.Bool(crashlytics.didCrashOnPreviousExecution())) } catch (error: Exception) { failure(error) }
        }
        Function("crash") { _ ->
            // Module calls may catch exceptions, so hand the crash straight to the uncaught
            // exception handler that Crashlytics installs.
            val thread = Thread.currentThread()
            thread.uncaughtExceptionHandler?.uncaughtException(thread, RuntimeException("Whisker Firebase Crashlytics test crash"))
            WhiskerValue.Null
        }
    }

    private fun stackTrace(report: Map<String, WhiskerValue>): Array<StackTraceElement> =
        (report["frames"] as? WhiskerValue.Array)?.value.orEmpty().mapNotNull { frame ->
            val fields = (frame as? WhiskerValue.Map)?.value ?: return@mapNotNull null
            val symbol = fields["symbol"]?.asString() ?: return@mapNotNull null
            val owner = symbol.substringBeforeLast("::", "rust")
            StackTraceElement(owner, symbol.substringAfterLast("::"), fields["file"]?.asString(), fields["line"]?.asInt()?.toInt() ?: -1)
        }.toTypedArray()

    private fun attempt(body: () -> Unit): WhiskerValue =
        try { body(); success(WhiskerValue.Null) } catch (error: Exception) { failure(error) }

    private fun success(value: WhiskerValue) = WhiskerValue.Map(mapOf("value" to value))
    private fun failure(error: Exception?): WhiskerValue {
        val code = when (error) {
            is IllegalArgumentException -> "invalid-argument"
            is IllegalStateException -> "app-not-configured"
            else -> "unknown"
        }
        return WhiskerValue.Map(mapOf("error" to WhiskerValue.Map(mapOf(
            "code" to WhiskerValue.Str(code),
            "message" to WhiskerValue.Str(error?.message ?: "Crashlytics operation failed"),
        ))))
    }
}

File: packages/whisker-firebase/crashlytics/ios/Sources/WhiskerFirebaseCrashlytics/FirebaseCrashlyticsModule.swift
import Foundation
import FirebaseCore
import FirebaseCrashlytics
import WhiskerModule

@WhiskerModule
public final class FirebaseCrashlyticsModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("FirebaseCrashlytics")
            Function("log") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let message = args.first?.asString else {
                    return Self.failure("invalid-argument", "Expected a message")
                }
                return Self.with { $0.log(message) }
            }
            Function("setUserId") { (args: [WhiskerValue]) -> WhiskerValue in
                Self.with { $0.setUserID(args.first?.asString) }
            }
            Function("setCustomKey") { (args: [WhiskerValue]) -> WhiskerValue in
                guard args.count == 2, let key = args[0].asString else {
                    return Self.failure("invalid-argument", "Expected a key and value")
                }
                let value: Any
                switch args[1] {
                case .string(let v): value = v
                case .int(let v): value = NSNumber(value: v)
                case .float(let v): value = NSNumber(value: v)
                case .bool(let v): value = NSNumber(value: v)
                default: return Self.failure("invalid-argument", "Unsupported custom key value")
                }
                return Self.with { $0.setCustomValue(value, forKey: key) }
            }
            Function("recordError") { (args: [WhiskerValue]) -> WhiskerValue in
                guard case .map(let report) = args.first, let name = report["name"]?.asString,
                      let reason = report["reason"]?.asString else {
                    return Self.failure("invalid-argument", "Expected an error report")
                }
                let model = ExceptionModel(name: name, reason: reason)
                if case .array(let frames) = report["frames"] {
                    model.stackTrace = frames.compactMap { frame in
                        guard case .map(let fields) = frame, let symbol = fields["symbol"]?.asString else { return nil }
                        return StackFrame(symbol: symbol, file: fields["file"]?.asString ?? "",
                                          line: Int(fields["line"]?.asInt ?? 0))
                    }
                }
                return Self.with { $0.record(exceptionModel: model) }
            }
            Function("recordPanic") { (args: [WhiskerValue]) -> WhiskerValue in
                guard case .map(let report) = args.first, let reason = report["reason"]?.asString else {
                    return Self.failure("invalid-argument", "Expected a panic report")
                }
                var text = reason
                if case .array(let frames) = report["frames"] {
                    for case .map(let frame) in frames {
                        text += "\n  at \(frame["symbol"]?.asString ?? "?")"
                        if let file = frame["file"]?.asString { text += " (\(file):\(frame["line"]?.asInt ?? 0))" }
                    }
                }
                // Crashlytics records through the main queue, which a panicking main thread never
                // reaches, but its terminate handler writes an uncaught NSException synchronously.
                // The exception is raised on a thread without Rust frames, which cannot unwind it.
                let exception = NSException(name: NSExceptionName("RustPanic"), reason: text)
                Thread.detachNewThread { exception.raise() }
                Thread.sleep(forTimeInterval: 5)
                return Self.success(.null)
            }
            Function("isCollectionEnabled") { (_: [WhiskerValue]) -> WhiskerValue in
                guard FirebaseApp.app() != nil else { return Self.notConfigured() }
                return Self.success(.bool(Crashlytics.crashlytics().isCrashlyticsCollectionEnabled()))
            }
            Function("setCollectionEnabled") { (args: [WhiskerValue]) -> WhiskerValue in
                guard let enabled = args.first?.asBool else {
                    return Self.failure("invalid-argument", "Expected a boolean")
                }
                return Self.with { $0.setCrashlyticsCollectionEnabled(enabled) }
            }
            AsyncFunction("checkForUnsentReports") { (_: [WhiskerValue], promise: WhiskerPromise) in
                guard FirebaseApp.app() != nil else { promise.resolve(Self.notConfigured()); return }
                Crashlytics.crashlytics().checkForUnsentReports { promise.resolve(Self.success(.bool($0))) }
            }
            Function("sendUnsentReports") { (_: [WhiskerValue]) -> WhiskerValue in
                Self.with { $0.sendUnsentReports() }
            }
            Function("deleteUnsentReports") { (_: [WhiskerValue]) -> WhiskerValue in
                Self.with { $0.deleteUnsentReports() }
            }
            Function("didCrashOnPreviousExecution") { (_: [WhiskerValue]) -> WhiskerValue in
                guard FirebaseApp.app() != nil else { return Self.notConfigured() }
                return Self.success(.bool(Crashlytics.crashlytics().didCrashDuringPreviousExecution()))
            }
            Function("crash") { (_: [WhiskerValue]) -> WhiskerValue in
                fatalError("Whisker Firebase Crashlytics test crash")
            }
        }
    }

    private static func with(_ body: (Crashlytics) -> Void) -> WhiskerValue {
        guard FirebaseApp.app() != nil else { return notConfigured() }
        body(Crashlytics.crashlytics())
        return success(.null)
    }

    private static func success(_ value: WhiskerValue) -> WhiskerValue { .map(["value": value]) }
    private static func failure(_ code: String, _ message: String) -> WhiskerValue {
        .map(["error": .map(["code": .string(code), "message": .string(message)])])
    }
    private static func notConfigured() -> WhiskerValue {
        failure("app-not-configured", "Call Crashlytics::instance() before using Crashlytics")
    }
}
